File: src/main/java/mnm/mods/tabbychat/util/ScaledDimension.java
package mnm.mods.tabbychat.util;

import net.minecraft.client.Minecraft;
import net.minecraft.util.math.MathHelper;

/**
 * Modified version of ScaledResolution. Supports arbitrary dimensions
 * 
 * @author Matthew
 */
public class ScaledDimension {

    private final double scaledWidthD;
    private final double scaledHeightD;
    private int scaledWidth;
    private int scaledHeight;
    private int scaleFactor;

    public ScaledDimension(int width, int height) {
        this.scaledWidth = width;
        this.scaledHeight = height;
        this.scaleFactor = 1;
        Minecraft mc = Minecraft.getMinecraft();
        boolean unicode = mc.isUnicode();
        int scale = mc.gameSettings.guiScale;

        if (scale == 0) {
            scale = 1000;
        }

        while (this.scaleFactor < scale && this.scaledWidth / (this.scaleFactor + 1) >= 320 && this.scaledHeight / (this.scaleFactor + 1) >= 240) {
            ++this.scaleFactor;
        }

        if (unicode && this.scaleFactor % 2 != 0 && this.scaleFactor != 1) {
            --this.scaleFactor;
        }

        this.scaledWidthD = (double) this.scaledWidth / (double) this.scaleFactor;
        this.scaledHeightD = (double) this.scaledHeight / (double) this.scaleFactor;
        this.scaledWidth = MathHelper.ceil(this.scaledWidthD);
        this.scaledHeight = MathHelper.ceil(this.scaledHeightD);
    }

    public int getScaledWidth() {
        return this.scaledWidth;
    }

    public int getScaledHeight() {
        return this.scaledHeight;
    }

    public double getScaledWidth_double() {
        return this.scaledWidthD;
    }

    public double getScaledHeight_double() {
        return this.scaledHeightD;
    }

    public int getScaleFactor() {
        return this.scaleFactor;
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/client/util/ScaledDimension.kt
package mnm.mods.tabbychat.client.util

import mnm.mods.tabbychat.util.mc
import net.minecraft.util.math.MathHelper

/**
 * Modified version of ScaledResolution. Supports arbitrary dimensions
 */
class ScaledDimension(width: Int, height: Int) {

    val scaledWidth_double: Double
    val scaledHeight_double: Double
    var scaledWidth: Int
        private set
    var scaledHeight: Int
        private set
    var scaleFactor: Int
        private set

    init {
        this.scaledWidth = width
        this.scaledHeight = height
        this.scaleFactor = 1
        val unicode = mc.forceUnicodeFont
        var scale = mc.gameSettings.guiScale

        if (scale == 0) {
            scale = 1000
        }

        while (this.scaleFactor < scale && this.scaledWidth / (this.scaleFactor + 1) >= 320 && this.scaledHeight / (this.scaleFactor + 1) >= 240) {
            ++this.scaleFactor
        }

        if (unicode && this.scaleFactor % 2 != 0 && this.scaleFactor != 1) {
            --this.scaleFactor
        }

        this.scaledWidth_double = this.scaledWidth.toDouble() / this.scaleFactor.toDouble()
        this.scaledHeight_double = this.scaledHeight.toDouble() / this.scaleFactor.toDouble()
        this.scaledWidth = MathHelper.ceil(this.scaledWidth_double)
        this.scaledHeight = MathHelper.ceil(this.scaledHeight_double)
    }
}
